### ISSUE TITLE
Change of second when updating models leads to different updated_at values in db vs model instance

### ISSUE THREAD

**Issue by graemlourens** [laravel/framework#35845]
- Laravel Version: v8.20.1
- PHP Version: 7.4.13
- Database Driver & Version: MariaDb 10.3.26

### Description:
We run a rather large test suite very frequently (up to 50 times a day, each pipeline approximately 4000 tests, via paratest running in parallel).

In the past year more and more frequently we had an issue that our test would fail due to unexpected updated_at values coming from our API, vs what was in the DB. Retrying the pipeline mostly fixed the issue, so we put it off to some kind of weird race condition in our code, but have just found out that its in laravel itsself and are now able to reproduce / explain it.

I guess this does not happen often in 'real life' as the case only happens if you create and update a model just at the end of a second, and the update code in laravel runs in exactly the time when the second 'changes' to the next one. I'll walk you through it with an example:

1) you create a new model and save it. Your model has a specific updated_at value, and it matches the one in the db (lets assume its xxxx-xx-xx 00:00:01)
2) you take that model and change a property and save it (which causes an update). Lets assume this code is still running inside this first second after midnight, Laravel in performUpdate() of Model.php will indeed run 'updateTimestamps' however as it will set the same updated_at value as the model already has, its not considered 'dirty'.
3) Laravel proceeds (still in performUpdate method) to determine which fields are 'dirty' and sees that one ore more are
4) Lets assume that right at this moment (either just before or just after $dirty = $this->getDirty(); is called) the second changes to xxxx-xx-xx 00:00:02
5) Laravel proceeds to run $this->setKeysForSaveQuery($query)->update($dirty) which ends up using 'addUpdatedAtColumn' method. In there it will fetch a fresh timestamp via freshTimestampString() which now will result in a DB call saving xxxx-xx-xx 00:00:02 as the updated_at value, but our model in the code still has xxxx-xx-xx 00:00:01

I am aware that this is a very rare condition, but the amount of tests we're running, its happening to us nearly every day at least once.

Sadly currently we're not quite sure how to easily fix this without causing rats tail issues. Always considering updated_at as 'dirty' would be an option, could however cause other issues..

### Steps To Reproduce:

Firstly to reproduce this condition you need to add a `sleep(2)` right before or right after `$dirty = $this->getDirty();` in performUpdate method of laravels Model.php.

Anywhere do run like this (with a model that has timestamps enabled and has ids)

```
    $myModel = new SomeModel();
    $myModel->myattribute = 'x';
    $myModel->save();

    $myModel->myattribute = 'y';
    $myModel->save();

    dump((string)$myModel->updated_at);

    $myFreshModel = SomeModel::findOrFail($myModel->id);

    dump((string)$myFreshModel->updated_at);
```

This will show you 2 different timestamps. The one from the fresh model will (depending on when you run it) be either 1-3 seconds difference.

**Comment by mfn** [comment]
If your tests rely on timestamps (never a good idea), you have to control them.

The easiest way within Laravel models to use `Carbon::setTestNow` (or `CarbonImmutable::setTestNow` if that's your thing).

I've similar scenario, though a couple thousand more tests a lot of APIs (rest, GraphQL) but ever since using this, never had such issues.

HTH

**Comment by graemlourens** [comment]
@mfn great tip, thank you, we'll look into if this is possible to encorporate into our test suites.

It however seems to me that this is a bug in laravel that would still need to be adressed, would you agree?

**Comment by mfn** [comment]
> It however seems to me that this is a bug in laravel that would still need to be adressed, would you agree?

I've been using the `::setTestNow()` since forever, since I basically never encountered this in tests.

I re-read your description once more:
> Firstly to reproduce this condition you need to add a sleep(2) right before or right after $dirty = $this->getDirty(); in performUpdate method of laravels Model.php.
…
> In there it will fetch a fresh timestamp via freshTimestampString() which now will result in a DB call saving xxxx-xx-xx 00:00:02 as the updated_at value, but our model in the code still has xxxx-xx-xx 00:00:01

Yes, I can reproduce this.

- `\Illuminate\Database\Eloquent\Model::performUpdate`
  calls
- `\Illuminate\Database\Eloquent\Builder::update`
  calls
- `\Illuminate\Database\Eloquent\Builder::addUpdatedAtColumn`
  which generates **the value** for the update_at column forthe **database query** only.
  But this is **never** replicated back on the **current** model.

For this to work, `\Illuminate\Database\Eloquent\Builder::update` would be needed to be changed to:
- get the column and it's updated timestamp value
- write it to the model property
- then pass it on to `\Illuminate\Database\Eloquent\Builder::update`

When I read the code correctly, `\Illuminate\Database\Eloquent\Builder::update` being also called from `\Illuminate\Database\Eloquent\Builder::increment` and `\Illuminate\Database\Eloquent\Builder::decrement` I think this also makes these methods suspects to this behaviour.

Whether it's a bug or not is probably debatable, the code does not read like "it's buggy", rather like it was simply implemented that way and that's that (just a guess). It's definitely a change in behaviour which will affect current code (albeit even implicitly) relying on this.

**Comment by driesvints** [comment]
Thanks for the report. Since it's such an edge case atm we're not going to focus on fixing this ourselves but we do appreciate if someone could send in a PR. Most likely in a next major release.

**Comment by graemlourens** [comment]
@driesvints I totally understand. Hope somebody else will chip in, and maybe we'll have some good idea coming up at some point. As its such a deep low-level code the chance to making things worse up is not insignificant.

**Comment by graemlourens** [comment]
@mfn I can confirm that setTestNow circumvents this issue, and as you pointed out a good best practise to rely on in any case.
We put this into the setUp of TestCase (which is called for all our tests) and merely hat do exempt / reset this for 2 tests (out of thousands) that actually relied on having real timestamps (very specific rate-limit / lock timeout contract tests)

For anybody who is battling with this issue therefore this is a workaround or actually a best practice solution as @mfn pointed out. I'd like like to see this resolved eventually as i still believe it is a bug, however very low priority.

**Comment by taylorotwell** [comment]
I think this we have a solid work around for this I will go ahead and close this as "wont fix" for now pending a more pressing reason for addressing it. Addressing it now may have unintended consequences.
